testdata/cheats/Cheats.sol

```diff
@@ -60,6 +60,10 @@ interface Cheats {
     function record() external;
     // Gets all accessed reads and write slot from a recording session, for a given address
     function accesses(address) external returns (bytes32[] memory reads, bytes32[] memory writes);
+    // Captures all the transaction logs
+    function captureLogs() external;
+    // Gets (and consumes) the next captured transaction log
+    function nextCapturedLog() external returns (bytes32[] memory topics,bytes memory data);
```

Comment:
What's the primary reasoning for this cheatcode vs. something like `vm.getLogs()` and popping/indexing off of the returned array?

Reply:
0. Hey @onbjerg I really appreciate your early review of this draft!
1. I'll do my best to be succint.
2. Disclaimer I'll do once: I have less than a week of experience in the marriage of Rust+Solidity 😅 !!! There are many features or patterns I ignore, and I'm most likely to propose stuff you already discussed and discarded. But well, there's only one way to start any jouney.

### Background

* Initially **I wanted something very simple™** and give the tester all the parsing work. After all, what I wanted as a feature is the ability to capture logs, leaving the UX problem as a future issue.

### Analysis

* What do we give to the tester if we invoke `vm.getLogs`?
   * Should it be a `bytes32[][]`. Maybe... But wait! `data` is not necessarily constained to 32 bytes!
   * So you have somehow an array of `(bytes32[], bytes)`. Can we do that? 🤔 
   * If we can't do above, can we just take the content of a `RawLog`, do a `(bytes32[], bytes)` tuple, `abi.encode` it into a `byte` field and then return a `byte[]` ? Seems cumbersome... Why don't we just go low-level and get the first log of the list in the simplest format we can afford?
* So in the end, I ended up thinking in UX!

### Proposal

* Having as cheatcodes `vm.captureLogs()` and `vm.nextCapturedLog()` within `forge`.
    * Then you just, in a FIFO fashion, **take** the first log from the list, leaving the next one in queue.
    * `vm.nextCapturedLog()` gives you back a `(bytes32[], bytes)`. The tester needs to do some quick decoding such as `abi.decode(data, (string))`.
    * Still working on how to work the Topic 0. For now its processing is to be at the tester discretion.
* Create later some wrapper at `forge-std` to `getLogs()`. This or these function(s) would be iterating over `nextCapturedLog`, taking care on indexing and decoding aspects.

### Wrapping up

My final goal is to be able to port the `Ethernaut` challenge into `forge`. I need the feature of capturing (and then reading those captured) logs. Let's take advantage we currently have some capacity on implementing this, and let's look towards the best solution 🙂

Reply:
Thanks for this info! I had the same thought as @onbjerg when looking at this. IMO I think the best UX would be something like this:
```solidity
interface Vm {
  // Records all transaction logs so they can be retrieved later.
  function recordLogs() external;

  // Returns an array of the recorded logs. `accessLog` is consistent with `accesses`, even
  // though semantics are a bit different. In general I don't feel strongly about all the cheatcode names 
  function accessLogs() external returns ((bytes32 memory topics, bytes memory data)[]);

  // Similar alternative.
  struct Log {
    bytes32[] topics;
    bytes data;
  }
  function logs() external returns (Log[] memory);
}
```

I don't personally see a big need for a cheatcode to access the next log, as opposed to just returning the full array and having the user loop through it or access a particular index. Providing the whole array instead is also nice because you if you know the index of an expected event it's cleaner to access it.

Reply:
Thanks for the write-up @bentobox19! Don't worry about being new to Rust + Solidity, we'll help you if you need :smile: Looks good so far.

I agree with @mds1's 2nd proposal here: it would be entirely feasible (and imo preferable) to return `Log` structs. The nice thing about having it return *all* logs it captured between `recordLogs` and `logs` (or `accessLogs`, not sure what the best name is) is that helper "cheatcodes" like `nextLog` can be built on top in Forge Std or other libraries :smile: 

So I would definitely propose that you:

1. Remove the `nextCapturedLog` cheatcode
2. Replace it with a `logs` or `accessLogs` (still not sure about naming) that returns `Log[]`

A note on behavior of the `logs` cheatcode: I think it should return all logs between the call to `recordLogs` and the call to `logs`. After this the internal buffer should be cleared - in other words, the combination of these two cheatcodes define the range in which the logs were emitted, if that makes sense?

Reply:
Thank you @mds1 and @onbjerg onbjerg for your prompt feedback: I am renaming to

```solidity
    // Record all the transaction logs
    function recordLogs() external;
    // Gets all the recorded logs
    function getRecordedLogs() external returns ((bytes32[] memory topics, bytes memory data)[]);
```

Let's pass on `getLogs()`, to avoid any potential confusion of the testers with the RPC method [eth_getLogs](https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_getlogs) 🤷‍♂️